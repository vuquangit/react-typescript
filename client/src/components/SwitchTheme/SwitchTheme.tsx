import React, { FC } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import { RootState } from '@/stores/rootReducer'
import { applyTheme } from '@/stores/Theme'
import Emoji from '@/components/Emoji'
import {
  SwitchWrapper,
  SwitchTrack,
  SwitchTrackLight,
  SwitchTrackDark,
  SwitchToggleContent,
  SwitchToggleThumb,
  SwitchScreenReader,
} from './SwitchTheme.styled'

const SwitchTheme: FC = () => {
  const dispatch = useDispatch()
  const themeMode = useSelector((state: RootState) => state.theme.themeMode)

  const onSwitchTheme = () => {
    themeMode === 'dark'
      ? dispatch(applyTheme('light'))
      : dispatch(applyTheme('dark'))
  }

  return (
    <SwitchWrapper onClick={onSwitchTheme} data-testid="switch-theme">
      <SwitchTrack>
        <SwitchTrackLight themeMode={themeMode} data-testid="track-light-mode">
          <SwitchToggleContent>
            <Emoji symbol="ðŸŒœ" label="middle-moon" />
          </SwitchToggleContent>
        </SwitchTrackLight>
        <SwitchTrackDark themeMode={themeMode} data-testid="track-dark-mode">
          <SwitchToggleContent>
            <Emoji symbol="ðŸŒž" label="sun" />
          </SwitchToggleContent>
        </SwitchTrackDark>
      </SwitchTrack>
      <SwitchToggleThumb
        themeMode={themeMode}
        data-testid="switch-toggle-thumb"
      />
      <SwitchScreenReader
        aria-label="Dark mode toggle"
        type="checkbox"
        checked={themeMode === 'light'}
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        onChange={() => {}}
      />
    </SwitchWrapper>
  )
}

export default SwitchTheme
