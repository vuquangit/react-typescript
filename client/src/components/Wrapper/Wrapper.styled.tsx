/**
 *
 * Styled for Wrapper
 *
 */

import styled from 'styled-components'
import {
  space,
  display,
  border,
  background,
  position,
  flexbox,
  layout,
} from 'styled-system'
import { WrapperProps } from './Wrapper.types'

export const WrapperComponent: React.FC<WrapperProps> = styled('div')<
  WrapperProps
>`
  ${space};
  ${display};
  ${border};
  ${background};
  ${position};
  ${flexbox};
  ${layout};
`
