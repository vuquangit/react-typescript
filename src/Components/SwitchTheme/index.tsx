import React, { FC } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import { darkTheme, lightTheme } from 'theme/theme'
import { RootState } from 'redux/rootReducer'
import { applyTheme } from 'redux/Theme'

import {
  SwitchWrapper,
  SwitchTrack,
  SwitchTrackLight,
  SwitchTrackDark,
  SwitchToggleContent,
  SwitchToggleThumb,
  SwitchScreenReader,
} from './SwitchTheme.styled'

const SwitchTheme: FC = () => {
  const dispatch = useDispatch()
  const isDarkMode = useSelector((state: RootState) => state.theme.isDarkMode)

  const onSwitchTheme = () => {
    isDarkMode
      ? dispatch(applyTheme(darkTheme))
      : dispatch(applyTheme(lightTheme))
  }

  return (
    <SwitchWrapper onClick={onSwitchTheme}>
      <SwitchTrack>
        <SwitchTrackLight isLightMode={!isDarkMode}>
          <SwitchToggleContent>ðŸŒœ</SwitchToggleContent>
        </SwitchTrackLight>
        <SwitchTrackDark isDarkMode={isDarkMode}>
          <SwitchToggleContent>ðŸŒž</SwitchToggleContent>
        </SwitchTrackDark>
      </SwitchTrack>
      <SwitchToggleThumb checked={isDarkMode} />
      <SwitchScreenReader
        aria-label="Dark mode toggle"
        type="checkbox"
        checked={isDarkMode}
      />
    </SwitchWrapper>
  )
}

export default SwitchTheme
