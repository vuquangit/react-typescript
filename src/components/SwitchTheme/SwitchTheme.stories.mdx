import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks'
import { action } from '@storybook/addon-actions'
import SwitchTheme from '.'
import AppWrapper from '@/test/supports/AppWrapper'
import Space from 'stories/components/styled-system/Space.mdx'
import Styles from 'stories/styles.mdx'

<Meta
  title="Story/Components/SwitchTheme"
  component={SwitchTheme}
  argTypes={{ onClick: { action: 'clicked' } }}
/>

<Styles />

# SwitchTheme

With `MDX` we can define a story for `SwitchTheme` right in the middle of our
Markdown documentation.

export const Template = (args) => (
  <AppWrapper themeMode="light">
    <SwitchTheme {...args} />
  </AppWrapper>
)

<Canvas>
  <Story name="Normal">{Template.bind({})}</Story>
</Canvas>

<div class="subheading">Props</div>

<Space />
