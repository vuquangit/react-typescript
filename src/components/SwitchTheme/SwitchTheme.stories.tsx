import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import AppWrapper from '@/test/supports/AppWrapper'

import SwitchTheme from '.'
import { BaseSwitchThemeProps } from './SwitchTheme.types'

export default {
  title: 'Story/Components/Component',
  component: SwitchTheme,
  argTypes: { onClick: { action: 'onSwitchTheme' } },
} as Meta

const Template: Story<BaseSwitchThemeProps> = (args) => (
  <AppWrapper themeMode="light">
    <SwitchTheme {...args}>Component story</SwitchTheme>
  </AppWrapper>
)

export const Normal = Template.bind({})
Normal.args = {}
