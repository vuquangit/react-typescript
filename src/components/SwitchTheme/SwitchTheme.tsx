import React, { FC } from 'react'
import { useQuery } from '@apollo/client'

import { GET_THEME_CURRENT } from '@/graphql/queries/getThemeCurrent'
import Emoji from '@/components/Emoji'
import { applyTheme } from '@/graphql/config/apollo-local-cache'

import { BaseSwitchThemeProps } from './SwitchTheme.types'
import {
  SwitchWrapper,
  SwitchTrack,
  SwitchTrackLight,
  SwitchTrackDark,
  SwitchToggleContent,
  SwitchToggleThumb,
  SwitchScreenReader,
} from './SwitchTheme.styled'

const SwitchTheme: FC<BaseSwitchThemeProps> = () => {
  const {
    data: { themeMode },
  } = useQuery(GET_THEME_CURRENT)

  const LIGHT = 'light'
  const DARK = 'dark'

  const onSwitchTheme = () => {
    if (themeMode === DARK) {
      applyTheme(LIGHT)
      localStorage.setItem('theme.themeMode', LIGHT)
    } else {
      applyTheme(DARK)
      localStorage.setItem('theme.themeMode', DARK)
    }
  }

  return (
    <SwitchWrapper onClick={onSwitchTheme} data-testid="switch-theme">
      <SwitchTrack>
        <SwitchTrackLight themeMode={themeMode} data-testid="track-light-mode">
          <SwitchToggleContent>
            <Emoji symbol="ðŸŒœ" label="middle-moon" />
          </SwitchToggleContent>
        </SwitchTrackLight>
        <SwitchTrackDark themeMode={themeMode} data-testid="track-dark-mode">
          <SwitchToggleContent>
            <Emoji symbol="ðŸŒž" label="sun" />
          </SwitchToggleContent>
        </SwitchTrackDark>
      </SwitchTrack>
      <SwitchToggleThumb
        themeMode={themeMode}
        data-testid="switch-toggle-thumb"
      />
      <SwitchScreenReader
        aria-label="Dark mode toggle"
        type="checkbox"
        data-testid="switch-theme-input"
        checked={themeMode === 'light'}
        readOnly
      />
    </SwitchWrapper>
  )
}

export default SwitchTheme
