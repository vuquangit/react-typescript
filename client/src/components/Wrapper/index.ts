/**
 *
 * Index for Wrapper
 *
 */

import Wrapper from './Wrapper'

export default Wrapper
