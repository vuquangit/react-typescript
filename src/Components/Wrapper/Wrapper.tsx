/**
 *
 * Wrapper
 *
 */

import React, { FC } from 'react'
import { WrapperProps } from './Wrapper.types'
import { WrapperComponent } from './Wrapper.styled'
import { ThemeProvider } from 'styled-components'
import * as theme from 'theme/theme'

const Wrapper: FC<WrapperProps> = (props: WrapperProps) => {
  return (
    <ThemeProvider theme={theme}>
      <WrapperComponent {...props} />
    </ThemeProvider>
  )
}

export default Wrapper
