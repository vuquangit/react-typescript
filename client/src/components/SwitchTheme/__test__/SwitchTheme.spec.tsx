/**
 *
 * Tests for Switch Theme
 *
 */

import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import SwitchTheme from '..'
import Wrapper from '@/test/supports/Wrapper'

describe('App SwitchTheme', () => {
  it('Test SwitchTheme with light mode', () => {
    const { container, getByTestId } = render(
      <Wrapper themeMode="light">
        <SwitchTheme />
      </Wrapper>
    )

    expect(container).toMatchSnapshot()
    expect(getByTestId('track-light-mode')).toHaveStyleRule('opacity', '1')
    expect(getByTestId('track-dark-mode')).toHaveStyleRule('opacity', '0')
    expect(getByTestId('switch-toggle-thumb')).toHaveStyleRule('left', '27px')

    fireEvent.click(getByTestId('switch-theme'))
    expect(getByTestId('track-light-mode')).toHaveStyleRule('opacity', '0')
    expect(getByTestId('track-dark-mode')).toHaveStyleRule('opacity', '1')
    expect(getByTestId('switch-toggle-thumb')).toHaveStyleRule('left', '1px')
  })

  it('Test SwitchTheme with dark mode', () => {
    const { container, getByTestId } = render(
      <Wrapper themeMode="dark">
        <SwitchTheme />
      </Wrapper>
    )

    expect(container).toMatchSnapshot()
    expect(getByTestId('track-light-mode')).toHaveStyleRule('opacity', '0')
    expect(getByTestId('track-dark-mode')).toHaveStyleRule('opacity', '1')
    expect(getByTestId('switch-toggle-thumb')).toHaveStyleRule('left', '1px')

    fireEvent.click(getByTestId('switch-theme'))
    expect(getByTestId('track-light-mode')).toHaveStyleRule('opacity', '1')
    expect(getByTestId('track-dark-mode')).toHaveStyleRule('opacity', '0')
    expect(getByTestId('switch-toggle-thumb')).toHaveStyleRule('left', '27px')
  })
})
