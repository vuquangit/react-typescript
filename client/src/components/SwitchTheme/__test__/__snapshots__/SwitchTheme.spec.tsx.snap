// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App SwitchTheme Test SwitchTheme with dark mode 1`] = `
.c0 {
  touch-action: pan-x;
  display: inline-block;
  position: relative;
  cursor: pointer;
  background-color: transparent;
  border: 0;
  padding: 0;
}

.c1 {
  width: 50px;
  height: 24px;
  padding: 0;
  border-radius: 30px;
  background-color: #4d4d4d;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
}

.c2 {
  position: absolute;
  width: 14px;
  height: 10px;
  top: 0;
  bottom: 0;
  margin-top: auto;
  margin-bottom: auto;
  line-height: 0;
  left: 8px;
  opacity: 0;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c4 {
  position: absolute;
  width: 10px;
  height: 10px;
  top: 0;
  bottom: 0;
  margin-top: auto;
  margin-bottom: auto;
  line-height: 0;
  right: 10px;
  opacity: 1;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 10px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  width: 10px;
}

.c5 {
  -webkit-transition: all 0.5s cubic-bezier(0.23,1,0.32,1) 0ms;
  transition: all 0.5s cubic-bezier(0.23,1,0.32,1) 0ms;
  position: absolute;
  top: 1px;
  left: 1px;
  width: 22px;
  height: 22px;
  border: 1px solid #4d4d4d;
  border-radius: 50%;
  background-color: #fafafa;
  box-sizing: border-box;
  -webkit-transition: all 0.25s ease;
  transition: all 0.25s ease;
  box-shadow: 0 0 2px 3px #0099e0;
}

.c6 {
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

<div>
  <div
    class="c0"
    data-testid="switch-theme"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        data-testid="track-light-mode"
      >
        <div
          class="c3"
        >
          <span
            aria-hidden="false"
            aria-label="middle-moon"
            class="emoji"
            role="img"
          >
            ðŸŒœ
          </span>
        </div>
      </div>
      <div
        class="c4"
        data-testid="track-dark-mode"
      >
        <div
          class="c3"
        >
          <span
            aria-hidden="false"
            aria-label="sun"
            class="emoji"
            role="img"
          >
            ðŸŒž
          </span>
        </div>
      </div>
    </div>
    <div
      class="c5"
      data-testid="switch-toggle-thumb"
    />
    <input
      aria-label="Dark mode toggle"
      class="c6"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`App SwitchTheme Test SwitchTheme with light mode 1`] = `
.c0 {
  touch-action: pan-x;
  display: inline-block;
  position: relative;
  cursor: pointer;
  background-color: transparent;
  border: 0;
  padding: 0;
}

.c1 {
  width: 50px;
  height: 24px;
  padding: 0;
  border-radius: 30px;
  background-color: #4d4d4d;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
}

.c2 {
  position: absolute;
  width: 14px;
  height: 10px;
  top: 0;
  bottom: 0;
  margin-top: auto;
  margin-bottom: auto;
  line-height: 0;
  left: 8px;
  opacity: 1;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c4 {
  position: absolute;
  width: 10px;
  height: 10px;
  top: 0;
  bottom: 0;
  margin-top: auto;
  margin-bottom: auto;
  line-height: 0;
  right: 10px;
  opacity: 0;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 10px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  width: 10px;
}

.c5 {
  -webkit-transition: all 0.5s cubic-bezier(0.23,1,0.32,1) 0ms;
  transition: all 0.5s cubic-bezier(0.23,1,0.32,1) 0ms;
  position: absolute;
  top: 1px;
  left: 27px;
  width: 22px;
  height: 22px;
  border: 1px solid #4d4d4d;
  border-radius: 50%;
  background-color: #fafafa;
  box-sizing: border-box;
  -webkit-transition: all 0.25s ease;
  transition: all 0.25s ease;
  box-shadow: 0 0 2px 3px #0099e0;
}

.c6 {
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

<div>
  <div
    class="c0"
    data-testid="switch-theme"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        data-testid="track-light-mode"
      >
        <div
          class="c3"
        >
          <span
            aria-hidden="false"
            aria-label="middle-moon"
            class="emoji"
            role="img"
          >
            ðŸŒœ
          </span>
        </div>
      </div>
      <div
        class="c4"
        data-testid="track-dark-mode"
      >
        <div
          class="c3"
        >
          <span
            aria-hidden="false"
            aria-label="sun"
            class="emoji"
            role="img"
          >
            ðŸŒž
          </span>
        </div>
      </div>
    </div>
    <div
      class="c5"
      data-testid="switch-toggle-thumb"
    />
    <input
      aria-label="Dark mode toggle"
      checked=""
      class="c6"
      type="checkbox"
    />
  </div>
</div>
`;
